### PR TITLE
[memory snapshots] _record_memory_history_legacy bug fix

### LINKED ISSUE #108208: `torch.cuda.memory._record_memory_history()`: Warning: Unsupported unwinding pattern
### 🐛 Describe the bug

Running on the latest `viable/strict`, calling just 
```
import torch
torch.cuda.memory._record_memory_history(
    True, trace_alloc_max_entries=100000, trace_alloc_record_context=True
)
```
raises a warning like:
```
[W unwind.cpp:194] Warning: Unsupported unwinding pattern: Address not in range (function unwinderFor)
```

Moreover, running a more end-to-end example (e.g. from https://zdevito.github.io/2022/12/09/memory-traces.html), the snapshot is not as expected:
```
import pickle
import torch
from torchvision.models import resnet18

torch.cuda.memory._record_memory_history(
    True, trace_alloc_max_entries=100000, trace_alloc_record_context=True
)

model = resnet18().cuda()
criterion = torch.nn.CrossEntropyLoss()
optimizer = torch.optim.SGD(model.parameters(), lr=0.001, momentum=0.9)

input = torch.rand(8, 3, 224, 224, device="cuda")
labels = torch.zeros(8, dtype=torch.long, device="cuda")

model.train()

outputs = model(input)
loss = criterion(outputs, labels)
loss.backward()
optimizer.step()

snapshot = torch.cuda.memory._snapshot()

with open(f"snapshot.pickle", "wb") as f:
    pickle.dump(snapshot, f)
```

Screenshot 2023-08-29 at 10 38 22 PM

I building on our PyTorch AWS cluster with:
```
TORCH_CUDA_ARCH_LIST=8.0; python setup.py develop
```

### Versions

PyTorch version: 2.1.0a0+gitad1ba6a
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.21.0
Libc version: glibc-2.31

Python version: 3.8.10 (default, May 19 2021, 18:05:58)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-1038-aws-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 11.6.112
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB
GPU 2: NVIDIA A100-SXM4-40GB
GPU 3: NVIDIA A100-SXM4-40GB
GPU 4: NVIDIA A100-SXM4-40GB
GPU 5: NVIDIA A100-SXM4-40GB
GPU 6: NVIDIA A100-SXM4-40GB
GPU 7: NVIDIA A100-SXM4-40GB

Nvidia driver version: 525.85.12
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          96
On-line CPU(s) list:             0-95
Thread(s) per core:              2
Core(s) per socket:              24
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           85
Model name:                      Intel(R) Xeon(R) Platinum 8275CL CPU @ 3.00GHz
Stepping:                        7
CPU MHz:                         2999.998
BogoMIPS:                        5999.99
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       1.5 MiB
L1i cache:                       1.5 MiB
L2 cache:                        48 MiB
L3 cache:                        71.5 MiB
NUMA node0 CPU(s):               0-23,48-71
NUMA node1 CPU(s):               24-47,72-95
Vulnerability Itlb multihit:     KVM: Mitigation: VMX unsupported
Vulnerability L1tf:              Mitigation; PTE Inversion
Vulnerability Mds:               Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown:          Mitigation; PTI
Vulnerability Mmio stale data:   Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves ida arat pku ospke

Versions of relevant libraries:
[pip3] bert-pytorch==0.0.1a4
[pip3] clip-anytorch==2.5.2
[pip3] CoCa-pytorch==0.0.7
[pip3] dalle2-pytorch==1.14.2
[pip3] ema-pytorch==0.2.3
[pip3] flake8==6.0.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.12.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] functorch==1.14.0a0+b71aa0b
[pip3] mypy==1.4.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.3
[pip3] open-clip-torch==2.20.0
[pip3] pytorch-transformers==1.2.0
[pip3] pytorch-triton==2.1.0+e6216047b8
[pip3] pytorch-warmup==0.1.1
[pip3] rotary-embedding-torch==0.2.3
[pip3] torch==2.1.0a0+gitad1ba6a
[pip3] torch-fidelity==0.3.0
[pip3] torch_geometric==2.4.0
[pip3] torch-struct==0.5
[pip3] torchaudio==2.1.0a0+c76d952
[pip3] torchdata==0.6.1
[pip3] torchdistx==0.3.0.dev0+cu111
[pip3] torchmetrics==0.11.4
[pip3] torchmultimodal==0.1.0a0
[pip3] torchrec-nightly==2023.6.13
[pip3] torchtext==0.16.0a0+60bea66
[pip3] torchvision==0.16.0a0+8768131
[pip3] torchviz==0.0.2
[pip3] vector-quantize-pytorch==1.6.11
[pip3] vit-pytorch==0.40.2
[conda] bert-pytorch              0.0.1a4                   dev_0    <develop>
[conda] blas                      1.0                         mkl  
[conda] clip-anytorch             2.5.2                    pypi_0    pypi
[conda] coca-pytorch              0.0.7                    pypi_0    pypi
[conda] dalle2-pytorch            1.14.2                   pypi_0    pypi
[conda] ema-pytorch               0.2.3                    pypi_0    pypi
[conda] functorch                 1.14.0a0+b71aa0b          pypi_0    pypi
[conda] magma-cuda110             2.5.2                         1    pytorch
[conda] magma-cuda111             2.5.2                         1    pytorch
[conda] mkl                       2021.3.0           h06a4308_520  
[conda] mkl-include               2021.3.0           h06a4308_520  
[conda] mkl-service               2.4.0                    pypi_0    pypi
[conda] mkl_fft                   1.3.0            py38h42c9631_2  
[conda] mkl_random                1.2.1            py38ha9443f7_2  
[conda] numpy                     1.19.0                   pypi_0    pypi
[conda] open-clip-torch           2.20.0                   pypi_0    pypi
[conda] pytorch-sphinx-theme      0.0.24                   pypi_0    pypi
[conda] pytorch-transformers      1.2.0                    pypi_0    pypi
[conda] pytorch-triton            2.1.0+9e3e10c5ed          pypi_0    pypi
[conda] pytorch-warmup            0.1.1                    pypi_0    pypi
[conda] rotary-embedding-torch    0.2.3                    pypi_0    pypi
[conda] torch                     2.0.1                    pypi_0    pypi
[conda] torch-fidelity            0.3.0                    pypi_0    pypi
[conda] torch-geometric           2.4.0                    pypi_0    pypi
[conda] torch-struct              0.5                      pypi_0    pypi
[conda] torchaudio                0.14.0a0+b326bc4          pypi_0    pypi
[conda] torchdata                 0.6.1                    pypi_0    pypi
[conda] torchdistx                0.3.0.dev0+cu111           dev_0    <develop>
[conda] torchmetrics              0.11.4                   pypi_0    pypi
[conda] torchmultimodal           0.1.0a0                   dev_0    <develop>
[conda] torchrec-nightly          2023.6.13                pypi_0    pypi
[conda] torchtext                 0.15.0a0+c047efe          pypi_0    pypi
[conda] torchvision               0.11.0a0+31d5336          pypi_0    pypi
[conda] torchviz                  0.0.2                    pypi_0    pypi
[conda] vector-quantize-pytorch   1.6.11                   pypi_0    pypi
[conda] vit-pytorch               0.40.2                   pypi_0    pypi

cc: @zdevito 

cc @ezyang @gchanan @zou3519 @ptrblck

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108260

The argment order for the legacy path got swapped in a recent patch.
Because there is still a blog post documenting the legacy interface
people are hitting this pathway.

This patch fixes #108208
I will also update the blog post to the new API so that people are
more likely to use the newer `_record_memory_history` API.

### CODE REVIEW

**Review by awgu — APPROVED** [review]
Thank you!
